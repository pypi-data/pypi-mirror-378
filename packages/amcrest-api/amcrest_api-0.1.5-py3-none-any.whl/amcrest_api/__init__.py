"""
API Wrapper for Amcrest V3.26

https://support.amcrest.com/hc/en-us/articles/17903073032973-Amcrest-HTTP-API-SDK
"""

from importlib import metadata as importlib_metadata


def get_version() -> str:
    """Version."""
    try:
        return importlib_metadata.version(__name__)
    except importlib_metadata.PackageNotFoundError:  # pragma: no cover
        return "unknown"


version: str = get_version()
