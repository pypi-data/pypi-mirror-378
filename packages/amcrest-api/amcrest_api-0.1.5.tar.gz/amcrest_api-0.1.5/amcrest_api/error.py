"""Amcrest API errors."""


class UnsupportedStreamSubtype(ValueError):
    """Exception raised when accessing unavailable substream."""
