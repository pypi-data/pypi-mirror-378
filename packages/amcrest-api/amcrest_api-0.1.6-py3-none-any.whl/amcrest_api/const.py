"""Constants."""

from enum import IntEnum, StrEnum


class ApiEndpoints(StrEnum):
    """API Endpoints."""

    CONFIG_MANAGER = "/cgi-bin/configManager.cgi"
    ENCODE = "/cgi-bin/encode.cgi"
    EVENT_MANAGER = "/cgi-bin/eventManager.cgi"
    GLOBAL = "/cgi-bin/global.cgi"
    MAGIC_BOX = "/cgi-bin/magicBox.cgi"
    PTZ = "/cgi-bin/ptz.cgi"
    REALTIME_STREAM = "/cam/realmonitor"
    SNAPSHOT = "/cgi-bin/snapshot.cgi"
    STORAGE_DEVICE = "/cgi-bin/storageDevice.cgi"


class StreamType(IntEnum):
    """Stream Types."""

    MAIN = 0
    SUBSTREAM1 = 1
    SUBSTREAM2 = 2


class StreamTypeName(StrEnum):
    """Stream Type Names."""

    MAIN = "Main"
    SUBSTREAM1 = "Sub Stream 1"
    SUBSTREAM2 = "Sub Stream 2"


STREAM_TYPE_DICT = {
    StreamType.MAIN: StreamTypeName.MAIN,
    StreamType.SUBSTREAM1: StreamTypeName.SUBSTREAM1,
    StreamType.SUBSTREAM2: StreamTypeName.SUBSTREAM2,
}
