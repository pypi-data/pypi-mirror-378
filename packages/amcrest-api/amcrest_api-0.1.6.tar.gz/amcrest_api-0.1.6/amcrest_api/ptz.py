"""PTZ Helpers."""

from __future__ import annotations

from dataclasses import KW_ONLY, dataclass, field
from enum import StrEnum
from typing import Any


@dataclass(frozen=True)
class PtzPresetData:
    """Data for PTZ preset."""

    index: int
    name: str


class Direction(StrEnum):
    """Directions."""

    UP = "UP"
    DOWN = "DOWN"
    LEFT = "LEFT"
    RIGHT = "RIGHT"
    LEFTUP = "LEFTUP"
    RIGHTUP = "RIGHTUP"
    LEFTDOWN = "LEFTDOWN"
    RIGHTDOWN = "RIGHTDOWN"


ALL_DIRECTIONS = set(Direction)


@dataclass(kw_only=True)
class PtzCapabilityData:
    """Data for PTZ capabilities."""

    pan: bool = False
    tilt: bool = False
    zoom: bool = False
    preset: bool = False
    tour: bool = False

    pan_min: float = 0.0
    pan_max: float = 0.0
    pan_speed_min: int = 1
    pan_speed_max: int = 1

    tilt_min: float = 0.0
    tilt_max: float = 0.0
    tilt_speed_min: int = 1
    tilt_speed_max: int = 1

    zoom_min: float = 1.0
    zoom_max: float = 1.0

    preset_min: int = 0
    preset_max: int = 0

    tour_min: int = 0
    tour_max: int = 0

    supported_directions: list[Direction] = field(default_factory=list[Direction])

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> PtzCapabilityData:
        """Create from a PTZ capabilities response."""
        caps: dict[str, Any] = response["caps"]

        def check_true(key: str) -> bool:
            if (cap_str := caps.get(key)) is not None:
                return cap_str.lower() == "true"
            return False

        ret = PtzCapabilityData(
            pan=check_true("Pan"),
            tilt=check_true("Tile"),  # the key is a typo in the API
            zoom=check_true("Zoom"),
            preset=check_true("Preset"),
            tour=check_true("Tour"),
        )
        if ret.pan:
            ret.pan_min = float(caps["PtzMotionRange"]["HorizontalAngle"][0])
            ret.pan_max = float(caps["PtzMotionRange"]["HorizontalAngle"][1])
            ret.pan_speed_min = int(caps["PanSpeedMin"])
            ret.pan_speed_max = int(caps["PanSpeedMax"])
            if ret.pan_min > ret.pan_max:
                ret.pan_min, ret.pan_max = ret.pan_max, ret.pan_min
        if ret.tilt:
            ret.tilt_min = float(caps["PtzMotionRange"]["VerticalAngle"][0])
            ret.tilt_max = float(caps["PtzMotionRange"]["VerticalAngle"][1])
            ret.pan_speed_min = int(caps["PanSpeedMin"])
            ret.pan_speed_max = int(caps["PanSpeedMax"])
            if ret.tilt_min > ret.tilt_max:
                ret.tilt_min, ret.tilt_max = ret.tilt_max, ret.tilt_min
        if ret.zoom:
            ret.zoom_min = float(caps["ZoomMin"])
            ret.zoom_max = float(caps["ZoomMax"])
        if ret.preset:
            ret.preset_min = int(caps["PresetMin"])
            ret.preset_max = int(caps["PresetMax"])
        if ret.tour:
            ret.tour_min = int(caps["TourMin"])
            ret.tour_max = int(caps["TourMax"])
        if ret.pan or ret.tilt:
            ret.supported_directions = sorted(
                list[Direction](
                    ALL_DIRECTIONS
                    ^ set[Direction](caps.get("UnSupportDirections", dict()).values())  # type: ignore
                )
            )
        return ret


@dataclass(kw_only=True)
class PtzStatusData:
    """Data for PTZ status."""

    action: str | None
    move_status: str
    zoom_status: str
    preset_id: int | None = None
    sequence: int | None = None
    position_pan: float
    position_tilt: float
    position_zoom: float
    pts: int | None
    utc: int | None

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> PtzStatusData:
        """Create from a PTZ Status response."""
        ptz_status_dict: dict[str, Any] = response["status"]  # expect nested dict
        if ptz_status_dict.get("PresetID") == "0":
            preset_id = None
        else:
            preset_id = int(ptz_status_dict["PresetID"])
        if ptz_status_dict.get("Sequence") == "0":
            sequence = None
        else:
            sequence = int(ptz_status_dict["Sequence"])
        return PtzStatusData(
            action=ptz_status_dict.get("Action"),
            move_status=ptz_status_dict["MoveStatus"],
            zoom_status=ptz_status_dict["ZoomStatus"],
            preset_id=preset_id,
            sequence=sequence,
            # Typo in API "Postion instead of Position" # codespell:ignore
            position_pan=float(ptz_status_dict["Postion"][0]),  # codespell:ignore
            position_tilt=float(ptz_status_dict["Postion"][1]),  # codespell:ignore
            position_zoom=float(ptz_status_dict["Postion"][2]),  # codespell:ignore
            pts=int(str(ptz_status_dict.get("PTS"))),
            utc=int(str(ptz_status_dict.get("UTC"))),
        )


@dataclass(kw_only=True)
class PtzRelativeMove:
    """Class for a relative move. All values are normalized from [-1,1]."""

    _caps: PtzCapabilityData | None = None
    _horizontal: float | None = None
    _vertical: float | None = None
    _zoom: float | None = None
    _channel: int = 1

    def __init__(
        self,
        *,
        caps: PtzCapabilityData | None = None,
        horizontal: float | None = None,
        vertical: float | None = None,
        zoom: float | None = None,
        channel: int = 1,
    ):
        self.caps = caps
        self.horizontal = horizontal
        self.vertical = vertical
        self.zoom = zoom
        self.channel = channel

    @property
    def caps(self) -> PtzCapabilityData | None:
        """PTZ capabilities."""
        return self._caps

    @caps.setter
    def caps(self, value: PtzCapabilityData | None):
        self._caps = value

    @property
    def horizontal(self) -> float | None:
        """Relative horizontal movement, [-1,1] inclusive"""
        return self._horizontal

    @horizontal.setter
    def horizontal(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._horizontal = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def vertical(self) -> float | None:
        """Relative vertical movement, [-1,1] inclusive"""
        return self._vertical

    @vertical.setter
    def vertical(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._vertical = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def zoom(self) -> float | None:
        """Relative zoom movement, [-1,1] inclusive"""
        return self._zoom

    @zoom.setter
    def zoom(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._zoom = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def channel(self) -> int:
        """PTZ channel."""
        return self._channel

    @channel.setter
    def channel(self, value: int) -> None:
        self._channel = int(value)

    def get_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary that can be passed as query parameters to the API Endpoint.

        arg1 is horizontal
        arg2 is vertical
        arg3 is zoom
        """
        ret: dict[str, Any] = {
            "action": "moveRelatively",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.horizontal is not None:
            ret["arg1"] = self.horizontal
        if self.vertical is not None:
            ret["arg2"] = self.vertical
        if self.zoom is not None:
            ret["arg3"] = self.zoom
        return ret


@dataclass
class PtzBasicMove:
    """Class for starting PTZ motion."""

    direction: Direction
    _: KW_ONLY
    caps: PtzCapabilityData | None = None
    speed: int = 1
    vertical_speed: int | None = None
    horizontal_speed: int | None = None
    channel: int = 1

    def get_start_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary that can be passed as query parameters to the API Endpoint.

        arg1 is vertical speed on combo moves
        arg2 is vertical or horizontal speed on single moves
        arg3 is 0
        """
        ret: dict[str, Any] = {
            "action": "start",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.direction in [
            Direction.LEFTDOWN,
            Direction.RIGHTDOWN,
            Direction.RIGHTUP,
            Direction.LEFTUP,
        ]:
            ret["arg1"] = self.vertical_speed or self.speed
            ret["arg2"] = self.horizontal_speed or self.speed

        elif self.direction in [
            Direction.LEFT,
            Direction.RIGHT,
            Direction.UP,
            Direction.DOWN,
        ]:
            ret["arg1"] = 0
            ret["arg2"] = self.speed
        ret["code"] = self.direction.title()
        ret["arg3"] = 0
        return ret

    def get_stop_query_dict(self) -> dict[str, Any]:
        """Get the stop query dict."""
        return {
            "action": "stop",
            "channel": self.channel,
            "code": self.direction.title(),
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }


@dataclass(kw_only=True)
class PtzAccuratePosition:
    """Class for Accurate position setting."""

    caps: PtzCapabilityData | None = None
    horizontal_position: float | None = None
    vertical_position: float | None = None
    zoom: float | None = None
    channel: int = 1

    def get_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary to be passed as query parameters to the Accurate Position API

        arg1 horizonatal position
        arg2 vertical position
        arg3 zoom
        """
        if self.caps is None:
            raise RuntimeError(
                "Cannot get query dict unless PTZ capabilities are assigned."
            )

        ret: dict[str, Any] = {
            "action": "moveAbsolutely",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.horizontal_position is not None and self.caps.pan:
            if self.caps.pan_min <= self.horizontal_position <= self.caps.pan_max:
                ret["arg1"] = (self.horizontal_position - 180.0) / 180.0
            else:
                raise ValueError(
                    f"Horizontal move degrees must be in the range [{self.caps.pan_min}, {self.caps.pan_max}, but it was {self.horizontal_position}]"  # noqa: E501
                )
        if self.vertical_position is not None and self.caps.tilt:
            if self.caps.tilt_min <= self.vertical_position <= self.caps.tilt_max:
                ret["arg2"] = self.vertical_position / 180.0
            else:
                raise ValueError(
                    f"Vertical move degrees must be in the range [{self.caps.tilt_min}, {self.caps.tilt_max}, but it was {self.vertical_position}]"  # noqa: E501
                )
        if self.zoom is not None and self.caps.zoom:
            if self.caps.zoom_min <= self.zoom <= self.caps.zoom_max:
                zoom_range = self.caps.zoom_max - self.caps.zoom_min
                zoom_mid = (self.caps.zoom_max + self.caps.zoom_min) / 2.0
                normed = (self.zoom - zoom_mid) / (zoom_range / 2)
                ret["arg3"] = normed
            else:
                raise ValueError(
                    f"Zoom must be in the range [{self.caps.zoom_min}, {self.caps.zoom_max}, but it was {self.zoom}]"  # noqa: E501
                )
        return ret
