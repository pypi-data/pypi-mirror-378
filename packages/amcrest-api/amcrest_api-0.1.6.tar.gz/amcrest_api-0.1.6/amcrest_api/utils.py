"""Utilities."""

import re
from typing import Any

import httpx


def parse_response(response: httpx.Response) -> bytes | dict:
    content_type: str = response.headers["content-type"]
    if "image/jpeg" in content_type:
        return response.content
    elif "text/plain" in content_type:
        return parse_key_value_response(response)
    elif "application/json" in content_type:
        return response.json()
    raise ValueError(f"Response does not contain supported content: {content_type}")


def parse_key_value_response(response: httpx.Response) -> dict[int | str, Any]:
    ret: dict[int | str, Any] = {}
    # The API can simply reply with OK too
    if response.text.strip() == "OK":
        return ret
    for line in response.iter_lines():
        keystr, val = line.split("=")
        keysraw = keystr.split(".")
        if keysraw[0] == "table":
            keysraw = keysraw[1:]
        keys: list[list[int | str]] = []
        for keyraw in keysraw:
            mainkey = keyraw.split("[")[0]  # get the key without indices
            indices = [int(x) for x in re.findall(r"\[(\d+)\]", keyraw)]
            keys.append([mainkey] + indices)
        # flatten the keys
        keysflat: list[int | str] = [x for xx in keys for x in xx]
        current = ret
        for keyflat in keysflat[:-1]:
            current = current.setdefault(keyflat, {})
        current[keysflat[-1]] = val

    return ret


def indexed_dict_to_list(indexed_dict: dict[int, Any]) -> list[Any]:
    """Converts a dict that could be a list, into a list."""
    return [indexed_dict[i] for i, key in enumerate(sorted(indexed_dict.keys()))]
