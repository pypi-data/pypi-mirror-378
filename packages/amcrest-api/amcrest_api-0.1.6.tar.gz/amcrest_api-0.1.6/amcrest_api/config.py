"""Config for information unlikely to change during a session."""

from dataclasses import dataclass
from typing import Any

from amcrest_api.const import StreamType, StreamTypeName

from .event import EventMessageType
from .ptz import PtzCapabilityData


@dataclass(frozen=True)
class Config:
    """A Config containing elements unlikely to change during the API Session."""

    device_type: str
    hardware_version: str
    machine_name: str
    max_extra_stream: StreamType
    network: dict[str, Any]
    ptz_capabilities: PtzCapabilityData
    serial_number: str
    session_physical_address: str
    supported_events: list[EventMessageType]
    supported_streams: dict[StreamType, StreamTypeName]
    software_version: str
