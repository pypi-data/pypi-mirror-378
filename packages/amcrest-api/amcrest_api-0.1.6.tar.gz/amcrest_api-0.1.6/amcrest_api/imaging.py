"""Classes for camra imaging."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum, IntFlag, StrEnum
from typing import Any

from .utils import indexed_dict_to_list


class Rotate90Flag(IntFlag):
    """Rotation flags."""

    NO_ROTATE = 0
    CLOCKWISE_90 = 1
    COUNTERCLOCKWISE_90 = 2


@dataclass(kw_only=True)
class VideoImageControl:
    """Video Image Control (Rotation/Flip)"""

    flip: bool = False
    freeze: bool = False
    mirror: bool = False
    rotate_90: Rotate90Flag = Rotate90Flag.NO_ROTATE
    stable: int = 0

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> list[VideoImageControl]:
        """Create from a Video Image Control response."""
        return [
            VideoImageControl(
                flip=control.get("Flip", "false") == "true",
                freeze=control.get("Freeze", "false") == "true",
                mirror=control.get("Mirror", "false") == "true",
                rotate_90=Rotate90Flag(int(control.get("Rotate90", "0"))),
                stable=int(control.get("Stable", "0")),
            )
            for control in indexed_dict_to_list(response["VideoImageControl"])
        ]


class VideoMode(StrEnum):
    """Selections for video modes."""

    COLOR = "Color"
    BRIGHTNESS = "Brightness"
    BLACK_WHITE = "BlackWhite"
    PHOTORESISTOR = "Photoresistor"
    GAIN = "Gain"


class Sensitivity(IntEnum):
    """Sensitivity settings."""

    LOW = 1
    MEDIUM = 2
    HIGH = 3


class ConfigNo(IntEnum):
    """Video Profile Config Index"""

    DAY = 0
    NIGHT = 1
    NORMAL = 2


CONFIG_NO_DICT = {
    ConfigNo.DAY: "Day",
    ConfigNo.NIGHT: "Night",
    ConfigNo.NORMAL: "Normal/General",
}


class VideoDayNightType(StrEnum):
    """Video Day/Night Types"""

    ELECTRON = "Electron"
    MECHANISM = "Mechanism"
    NIGHT_ICR = "NightICR"
    AUTO = "Auto"


@dataclass(kw_only=True)
class VideoDayNight:
    """Video Day/Night Settings"""

    delay_seconds: int
    mode: VideoMode
    sensitivity: Sensitivity
    type: VideoDayNightType

    @staticmethod
    def create_from_response(
        response: dict[str, Any],
    ) -> list[list[VideoDayNight]]:
        """Create from API response."""
        return [
            [
                VideoDayNight(
                    delay_seconds=int(config["Delay"]),
                    mode=VideoMode(config["Mode"]),
                    sensitivity=Sensitivity(int(config["Sensitivity"])),
                    type=VideoDayNightType(config["Type"]),
                )
                for config in indexed_dict_to_list(video_in_day_night)
            ]
            for video_in_day_night in indexed_dict_to_list(response["VideoInDayNight"])
        ]


class LightingMode(StrEnum):
    """Lighting Modes."""

    MANUAL = "Manual"
    AUTO = "Auto"
    OFF = "Off"


@dataclass(kw_only=True)
class Light:
    """Light."""

    angle: int = 50
    light: int


@dataclass(kw_only=True)
class Lighting:
    """Lighting config."""

    mode: LightingMode
    correction: int | None = None
    # far_light: int
    middle_light: Light | None
    # near_light: int
    light_type: str | None = None
    sensitivity: Sensitivity | None = None

    @staticmethod
    def create_from_response(
        response: dict[str, Any],
    ) -> list[list[list[Lighting]]]:
        """Create from API response."""
        return [
            [
                [
                    Lighting(
                        mode=light["Mode"],
                        correction=light["Correction"],
                        # far_light=int(light["FarLight"][0]["Light"]),
                        middle_light=Light(
                            light=int(light["MiddleLight"][0]["Light"]),
                            angle=int(light["MiddleLight"][0].get("Angle", 50)),
                        ),
                        light_type=light.get("LightType"),
                        # near_light=int(light["NearLight"][0]["Light"]),
                        sensitivity=Sensitivity(int(light["Sensitive"])),
                    )
                    for light in indexed_dict_to_list(config)
                ]
                for config in indexed_dict_to_list(lighting)
            ]
            for lighting in indexed_dict_to_list(response["Lighting_V2"])
        ]
