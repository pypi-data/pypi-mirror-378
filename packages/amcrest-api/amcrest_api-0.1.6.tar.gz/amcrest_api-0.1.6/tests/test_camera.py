"""Tests the camera"""

from datetime import datetime
from typing import TYPE_CHECKING

import pytest
import yarl
from pytest_httpserver import HTTPServer, RequestMatcher

from amcrest_api.camera import Camera
from amcrest_api.const import ApiEndpoints, StreamType
from amcrest_api.error import UnsupportedStreamSubtype

if TYPE_CHECKING:
    from amcrest_api.ptz import PtzCapabilityData, PtzStatusData


async def test_serial_number(camera: Camera) -> None:
    """Test serial number."""
    assert await camera.async_serial_number == "AMC00123456789ABCDEF"


async def test_lighting(camera: Camera, snapshot) -> None:
    """Test lighting."""
    assert await camera.async_lighting_config == snapshot


async def test_get_privacy_mode_on(camera: Camera) -> None:
    """Test Privacy mode, fixture was saved with it 'on'."""
    assert await camera.async_get_privacy_mode_on()


async def test_get_smart_track_on(camera: Camera) -> None:
    """Test Smart track, fixture was saved with it 'off'."""
    assert not await camera.async_get_smart_track_on()


async def test_read_fixed_config(camera: Camera, snapshot) -> None:
    """Test get physical config parameters unexpected to change."""
    assert await camera.async_get_fixed_config() == snapshot


async def test_read_ptz_config(camera: Camera, snapshot) -> None:
    """Test get PTZ config."""
    assert await camera.async_ptz_preset_info == snapshot


async def test_read_ptz_config_empty_presets(camera_no_ptz_presets: Camera) -> None:
    """Test get PTZ config when no presets are set."""
    presets = await camera_no_ptz_presets.async_ptz_preset_info
    assert len(presets) == 0


async def test_read_ptz_config_no_caps(camera_no_ptz_caps: Camera) -> None:
    """Test get PTZ config when no Capabilities."""

    caps = await camera_no_ptz_caps.async_ptz_capabilities
    assert not caps.pan
    assert not caps.tilt
    assert not caps.zoom
    assert not caps.tour


async def test_get_rtsp_url(camera: Camera) -> None:
    """Terst getting the RTSP URL"""
    url = yarl.URL(await camera.async_get_rtsp_url())
    assert str(url.host) == "localhost"
    assert str(url.scheme) == "rtsp"
    assert url.user
    assert url.password
    url = yarl.URL(await camera.async_get_rtsp_url(subtype=StreamType.SUBSTREAM1))
    assert str(url.host) == "localhost"
    assert str(url.scheme) == "rtsp"
    assert url.user
    assert url.password
    assert url.query["subtype"] == "1"
    with pytest.raises(UnsupportedStreamSubtype):
        await camera.async_get_rtsp_url(subtype=StreamType.SUBSTREAM2)


async def test_get_ptz_status(camera: Camera) -> None:
    """Test getting PTZ status."""
    status: PtzStatusData = await camera.async_ptz_status
    assert status.position_pan == 242.7
    assert status.position_tilt == 9.6
    assert status.position_zoom == 1.0
    assert status.preset_id is None


async def test_get_ptz_capabilities(
    camera: Camera, mock_camera_server: HTTPServer, snapshot
) -> None:
    """Test getting PTZ capabilities."""
    caps: PtzCapabilityData = await camera.async_ptz_capabilities
    assert set(caps.supported_directions) == {"LEFT", "UP", "DOWN", "RIGHT"}
    assert caps.pan_min == 1
    assert caps.pan_max == 354
    assert caps.tilt_min == -4
    assert caps.tilt_max == 79
    assert caps == snapshot

    # and it's cached
    mock_camera_server.assert_request_made(
        RequestMatcher(uri=ApiEndpoints.PTZ), count=1
    )

    _ = await camera.async_ptz_capabilities

    mock_camera_server.assert_request_made(
        RequestMatcher(uri=ApiEndpoints.PTZ), count=1
    )


async def test_get_storage_info_names(camera: Camera) -> None:
    """Test getting storage info names."""
    storage_info_names = await camera.async_storage_list
    assert len(storage_info_names) == 3
    assert "/dev/sda0" in storage_info_names
    assert "/dev/sda1" in storage_info_names
    assert "/dev/mmc0" in storage_info_names


async def test_get_storage_info_names_no_storage(camera_no_storage: Camera) -> None:
    """Test getting storage info names."""
    storage_info_names = await camera_no_storage.async_storage_list
    assert len(storage_info_names) == 0
    storage_infos = await camera_no_storage.async_storage_info
    assert len(storage_infos) == 0


async def test_get_storage_info(camera: Camera, snapshot) -> None:
    """Test getting storage info."""
    storage_infos = await camera.async_storage_info
    assert len(storage_infos) == 1
    assert not storage_infos[0].cant_hot_plug
    assert storage_infos[0].total_bytes == 997703680
    assert storage_infos == snapshot


async def test_get_video_image_control(camera: Camera, snapshot) -> None:
    """Test getting video image control."""
    video_image_control = await camera.async_video_image_control
    assert video_image_control == snapshot


async def test_get_video_in_day_night(camera: Camera, snapshot) -> None:
    """Test getting video in day night."""
    video_in_day_night = await camera.async_get_video_in_day_night()
    assert video_in_day_night == snapshot


async def test_get_current_time(camera: Camera) -> None:
    """Test getting current time."""
    current_time = await camera.async_get_current_time()
    assert current_time == datetime(2011, 7, 3, 21, 2, 32)
