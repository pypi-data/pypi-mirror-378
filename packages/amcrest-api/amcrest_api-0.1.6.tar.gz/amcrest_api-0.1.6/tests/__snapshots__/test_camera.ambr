# serializer version: 1
# name: test_get_ptz_capabilities
  PtzCapabilityData(pan=True, tilt=True, zoom=False, preset=True, tour=True, pan_min=1.0, pan_max=354.0, pan_speed_min=1, pan_speed_max=255, tilt_min=-4.0, tilt_max=79.0, tilt_speed_min=1, tilt_speed_max=1, zoom_min=1.0, zoom_max=1.0, preset_min=1, preset_max=25, tour_min=1, tour_max=8, supported_directions=[<Direction.DOWN: 'DOWN'>, <Direction.LEFT: 'LEFT'>, <Direction.RIGHT: 'RIGHT'>, <Direction.UP: 'UP'>])
# ---
# name: test_get_storage_info
  list([
    StorageDeviceInfo(is_error=False, path='/mnt/sd', type='ReadWrite', cant_hot_plug=False, total_bytes=997703680, used_bytes=26353664, life_percent=0.0, state='Success', sd_encrypt_flag=0, health_data_flag=0, name='/dev/mmc0'),
  ])
# ---
# name: test_get_video_image_control
  list([
    VideoImageControl(flip=False, freeze=False, mirror=False, rotate_90=<Rotate90Flag.COUNTERCLOCKWISE_90: 2>, stable=0),
  ])
# ---
# name: test_get_video_in_day_night
  list([
    list([
      VideoDayNight(delay_seconds=6, mode=<VideoMode.BRIGHTNESS: 'Brightness'>, sensitivity=<Sensitivity.MEDIUM: 2>, type=<VideoDayNightType.MECHANISM: 'Mechanism'>),
      VideoDayNight(delay_seconds=6, mode=<VideoMode.BRIGHTNESS: 'Brightness'>, sensitivity=<Sensitivity.MEDIUM: 2>, type=<VideoDayNightType.MECHANISM: 'Mechanism'>),
      VideoDayNight(delay_seconds=6, mode=<VideoMode.BRIGHTNESS: 'Brightness'>, sensitivity=<Sensitivity.MEDIUM: 2>, type=<VideoDayNightType.MECHANISM: 'Mechanism'>),
    ]),
  ])
# ---
# name: test_lighting
  list([
    list([
      list([
        Lighting(mode='Manual', correction='50', middle_light=Light(angle=50, light=100), light_type='InfraredLight', sensitivity=<Sensitivity.HIGH: 3>),
      ]),
      list([
        Lighting(mode='Manual', correction='50', middle_light=Light(angle=50, light=100), light_type='InfraredLight', sensitivity=<Sensitivity.HIGH: 3>),
      ]),
      list([
        Lighting(mode='Auto', correction='50', middle_light=Light(angle=50, light=100), light_type='InfraredLight', sensitivity=<Sensitivity.HIGH: 3>),
      ]),
    ]),
  ])
# ---
# name: test_read_fixed_config
  Config(device_type='IP4M-1041B', hardware_version='1.00', machine_name='AMC0TEST_CAMERA', max_extra_stream=1, network={'DefaultInterface': 'eth0', 'Domain': 'dauha', 'Hostname': 'IPC', 'eth0': {'DefaultGateway': '10.0.0.1', 'DhcpEnable': 'true', 'DnsAutoGet': 'false', 'DnsServers': {0: '8.8.8.8', 1: '8.8.4.4'}, 'EnableDhcpReservedIP': 'false', 'IPAddress': 'localhost', 'MTU': '1500', 'PhysicalAddress': 'a0:60:32:ff:ff:ff', 'SubnetMask': '255.0.0.0'}, 'eth2': {'DefaultGateway': '10.0.0.0', 'DhcpEnable': 'true', 'DnsAutoGet': 'false', 'DnsServers': {0: '8.8.8.8', 1: '8.8.4.4'}, 'EnableDhcpReservedIP': 'false', 'IPAddress': '', 'MTU': '1500', 'PhysicalAddress': 'a0:60:32:ff:ff:ff', 'SubnetMask': '255.0.0.0'}}, ptz_capabilities=PtzCapabilityData(pan=True, tilt=True, zoom=False, preset=True, tour=True, pan_min=1.0, pan_max=354.0, pan_speed_min=1, pan_speed_max=255, tilt_min=-4.0, tilt_max=79.0, tilt_speed_min=1, tilt_speed_max=1, zoom_min=1.0, zoom_max=1.0, preset_min=1, preset_max=25, tour_min=1, tour_max=8, supported_directions=[<Direction.DOWN: 'DOWN'>, <Direction.LEFT: 'LEFT'>, <Direction.RIGHT: 'RIGHT'>, <Direction.UP: 'UP'>]), serial_number='AMC00123456789ABCDEF', session_physical_address='a0:60:32:ff:ff:ff', supported_events=['AudioAnomaly', 'AudioMutation', 'LoginFailure', 'SafetyAbnormal', 'StorageFailure', 'StorageLowSpace', 'StorageNotExist', 'VideoBlind', 'VideoMotion', 'FaceDetection'], supported_streams={<StreamType.MAIN: 0>: <StreamTypeName.MAIN: 'Main'>, <StreamType.SUBSTREAM1: 1>: <StreamTypeName.SUBSTREAM1: 'Sub Stream 1'>}, software_version='2.800.00AC006.0.R,build:2023-10-27')
# ---
# name: test_read_ptz_config
  list([
    PtzPresetData(index='1', name='Preset1'),
    PtzPresetData(index='2', name='Preset2'),
    PtzPresetData(index='3', name='Preset3'),
    PtzPresetData(index='5', name='Preset5'),
  ])
# ---
